lib/seahorn/BvOpSem2.cc

```diff
@@ -3182,6 +3220,61 @@ Optional<APInt> Bv2OpSem::agg(Type *aggTy,
   return res;
 }
 
+void Bv2OpSem::initCrabAnalysis(const llvm::Module &M) {
+  // Get seadsa -- pointer analysis
+  auto &dsa_pass = m_pass.getAnalysis<seadsa::ShadowMemPass>().getShadowMem();
+  auto &dsa = dsa_pass.getDsaAnalysis();
+
+  // XXX: use of legacy operational semantics
+  auto &tli = m_pass.getAnalysis<TargetLibraryInfoWrapperPass>();
+
+  std::unique_ptr<clam::HeapAbstraction> heap_abs(
+      new clam::SeaDsaHeapAbstraction(M, dsa));
+
+  // -- Set parameters for CFG
+  clam::CrabBuilderParams cfg_builder_params;
+  LOG("opsem-rng", cfg_builder_params.print_cfg = true;);
+  cfg_builder_params.setPrecision(clam::CrabBuilderPrecision::MEM);
+  cfg_builder_params.interprocedural = true;
+
+  m_cfg_builder_man.reset(new clam::CrabBuilderManager(cfg_builder_params, tli,
+                                                       std::move(heap_abs)));
+  // -- Initialize crab for solving ranges
+  m_crab_rng_solver.reset(new clam::InterGlobalClam(M, *m_cfg_builder_man));
+}
+
+void Bv2OpSem::runCrabAnalysis() {
+  /// Set Crab parameters
+  clam::AnalysisParams aparams;
+  aparams.dom = clam::CrabDomain::INTERVALS;
+  aparams.run_inter = true;
+  aparams.check = clam::CheckerKind::NOCHECKS;
+  /// Run the Crab analysis
+  clam::ClamGlobalAnalysis::abs_dom_map_t assumptions;
+  LOG("opsem-rng", aparams.print_invars = true;);
+  m_crab_rng_solver->analyze(aparams, assumptions);
+}
+
+const llvm::ConstantRange
+Bv2OpSem::getCrabInstRng(const llvm::Instruction &I) const {
+  if (!m_crab_rng_solver)
+    return llvm::ConstantRange::getFull(64);
+  return m_crab_rng_solver->range(I);
+}
+
+void Bv2OpSem::runLVIAnalysis(llvm::Function &F) {
+  m_lvi = &m_pass.getAnalysis<LazyValueInfoWrapperPass>(F);
```

Comment:
this seems very bad. There is one global place to store `lvi` but there are many `lvi` per function.

Something is very wrong here